README.md
# gktool

Cli tool to generate secure key

⚠ **_It's new version of [gpwd](https://github.com/Ulrichinho/gpwd) except that It doesn't call API because it integrates an algorithme of generate of keys and some functions to ckeck strength of generation_**

## Install

**_Golang 1.17 must be install to add gktool cli_**

Go ahead compile it yourself :

```sh
go install github.com/Ulrichinho/gktool@latest
```

**_It's compatible with macOs, linux and Windows_**

## Usage

```text
gktool [global options]
```

### Global options

```text
GLOBAL OPTIONS:
   --length value, -l value    define length of key (default: 16)
   --quantity value, -q value  define quantity of key (default: 1)
   --no-upper                  define if you don't want upper chars (default: false)
   --no-lower                  define if you don't want lower chars (default: false)
   --no-number                 define if you don't want number chars (default: false)
   --no-symbol                 define if you don't want symbol chars (default: false)
   --export, -e                export generate key(s) in file (default: false)
   --help, -h                  show help (default: false)
   --version, -V               print only the version (default: false)

```

## Version

To see version :

```text
gktool [-V | --version]
```

see [`CHANGELOG`](./CHANGELOG.md) for more details about versions

## Help

To have more help about this command type :

```text
gktool [-h | --help]
```

## License

The Apache License (APACHE) - see [`LICENSE`](./LICENSE.md) for more details


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [v1.0.0] - 2021-12-23

This tool generate key(s) with possibility to change length, if have upper chars, lower chars, number chars and symbol chars. And It's possible to export generate key in 'password.txt'.

### Added

### Changed

### Fixed

install.sh
#!/usr/bin/sh

# Check go install
if ! command -v go &> /dev/null
then
    echo -e "[\033[5;38;5;160mERROR\033[0m]\033[38;5;160m Go isn't installed\033[0m"
    echo -e "[\033[5;38;5;114mHELP\033[0m] Refer to --> https://go.dev/doc/install"
    exit 1
fi

# Check go version
if [[ $(go version) == "go version go1.17."* ]]
then
    echo -e "[\033[5;38;5;75mINFO\033[0m] You have a go version >= 1.17"
else 
    echo -e "[\033[5;38;5;160mERROR\033[0m]\033[38;5;160m Go version isn't compatible\033[0m"
    echo -e "[\033[5;38;5;114mHELP\033[0m] Download new version or version >= 1.17. Refer to --> https://go.dev/doc/install"
    exit 1
fi

# Build gktool
cd $GOPATH
go build
if test $? -eq 0
then
    echo -e "[\033[5;38;5;34mSUCCESS\033[0m] BUILD bin"
fi

# Move build in bin path
sudo mv gktool /usr/bin
if [[ $? -eq 0 ]]
then
    echo -e "[\033[5;38;5;34mSUCCESS\033[0m] MV gktool in bin path"
fi


main.go
/*
Copyright © 2021 Ulrichinho <grolhier.u@gmail.com>
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package main

import (
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"regexp"
	"time"

	"github.com/urfave/cli/v2"
)

func getChars() []string {
	return []string{
		"ABCDEFGHIJKLMNOPQRSTUVWXYZ",
		"abcdefghijklmnopqrstuvwxyz",
		"0123456789",
		"!\"#$%&'()*+,-.:;<=>?@[]^_"}
}

type Key struct {
	length   int
	isUpper  bool
	isLower  bool
	isNumber bool
	isSymbol bool
	value    string
	strength float64
}

func isUpper(noUpper bool) bool {
	return !noUpper
}

func isLower(noLower bool) bool {
	return !noLower
}

func isNumber(noNumber bool) bool {
	return !noNumber
}

func isSymbol(noSymbol bool) bool {
	return !noSymbol
}

func NewKey(length int, noUpper bool, noLower bool, noNumber bool, noSymbol bool) Key {
	k := Key{}
	k.length = length
	k.isUpper = isUpper(noUpper)
	k.isLower = isLower(noLower)
	k.isNumber = isNumber(noNumber)
	k.isSymbol = isSymbol(noSymbol)
	k.value = generate(k)
	k.strength = calculateEntropy(k.value, k.length)
	return k
}

func getRandomChars(chars string) string {
	i := rand.Intn(len(chars))
	return string(chars[i])
}

func generate(k Key) string {
	if !k.isUpper && !k.isLower && !k.isNumber && !k.isSymbol {
		fmt.Println("[\033[5;38;5;160mERROR\033[0m] Cannot create key with no chars!")
		os.Exit(1)
	}

	for len(k.value) != k.length {
		gindex := rand.Intn(4)
		if gindex == 0 && k.isUpper {
			k.value += getRandomChars(getChars()[0])
		} else if gindex == 1 && k.isLower {
			k.value += getRandomChars(getChars()[1])
		} else if gindex == 2 && k.isNumber {
			k.value += getRandomChars(getChars()[2])
		} else if gindex == 3 && k.isSymbol {
			k.value += getRandomChars(getChars()[3])
		}
	}
	return k.value
}

func calculateStrength(key string) int {
	strength := 0
	patterns := []*regexp.Regexp{regexp.MustCompile(`^.*[a-z]`), regexp.MustCompile(`^.*[A-Z]`), regexp.MustCompile(`^.*[0-9]`), regexp.MustCompile(`/^.*[!\"#$%&'()*+,-.:;<=>?@[]^_"]/`)}
	for i := 0; i < len(patterns); i++ {
		if patterns[i].MatchString(key) {
			switch i {
			case 0:
				strength += 26
			case 1:
				strength += 26
			case 2:
				strength += 10
			case 3:
				strength += 32
			default:
				strength += 0
			}
		}
	}
	return strength
}

func calculateEntropy(key string, l int) float64 {
	if l == 0 {
		return 0
	}
	strength := calculateStrength(key)
	return float64(l) * math.Log2(float64(strength))
}

type StrengthColor string

const (
	LOW    StrengthColor = "\033[38;5;124m"
	MEDIUM StrengthColor = "\033[38;5;202m"
	STRONG StrengthColor = "\033[38;5;28m"
)

func exportKeyInFile(k []Key) {
	f, err := os.OpenFile("password.txt", os.O_CREATE|os.O_WRONLY, 0600)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	for i := 0; i < len(k); i++ {
		if _, err := f.WriteString(k[i].value + "\n"); err != nil {
			log.Fatal(err)
		}
	}
}

func main() {
	// Flags
	var length int
	var quantity int
	var noUpper bool
	var noLower bool
	var noNumber bool
	var noSymbol bool
	var export bool

	var color StrengthColor

	app := &cli.App{
		Name:    "gktool",
		Usage:   "generate key",
		Version: "v1.0.0",
		Flags: []cli.Flag{
			&cli.IntFlag{
				Name:        "length",
				Aliases:     []string{"l"},
				Usage:       "define length of key",
				Value:       16,
				Destination: &length,
			},
			&cli.IntFlag{
				Name:        "quantity",
				Aliases:     []string{"q"},
				Usage:       "define quantity of key",
				Value:       1,
				Destination: &quantity,
			},
			&cli.BoolFlag{
				Name:        "no-upper",
				Usage:       "define if you don't want upper chars",
				Value:       false,
				Destination: &noUpper,
			},
			&cli.BoolFlag{
				Name:        "no-lower",
				Usage:       "define if you don't want lower chars",
				Value:       false,
				Destination: &noLower,
			},
			&cli.BoolFlag{
				Name:        "no-number",
				Usage:       "define if you don't want number chars",
				Value:       false,
				Destination: &noNumber,
			},
			&cli.BoolFlag{
				Name:        "no-symbol",
				Usage:       "define if you don't want symbol chars",
				Value:       false,
				Destination: &noSymbol,
			},
			&cli.BoolFlag{
				Name:        "export",
				Aliases:     []string{"e"},
				Usage:       "export generate key(s) in file",
				Value:       false,
				Destination: &export,
			},
		},
		Action: func(c *cli.Context) error {
			// ASCII art
			fmt.Printf("\n  \033[48;0;36m██████\033[37m╗ \033[48;0;36m██\033[37m╗  \033[48;0;36m██\033[37m╗\033[38;5;125m████████\033[37m╗ \033[38;5;125m██████\033[37m╗  \033[38;5;125m██████\033[37m╗ \033[38;5;125m██\033[37m╗     \n")
			fmt.Printf(" \033[48;0;36m██\033[37m╔════╝ \033[48;0;36m██\033[37m║ \033[48;0;36m██\033[37m╔╝╚══\033[38;5;125m██\033[37m╔══╝\033[38;5;125m██\033[37m╔═══\033[38;5;125m██\033[37m╗\033[38;5;125m██\033[37m╔═══\033[38;5;125m██\033[37m╗\033[38;5;125m██\033[37m║     \n")
			fmt.Printf(" \033[48;0;36m██\033[37m║  \033[48;0;36m███\033[37m╗\033[48;0;36m█████\033[37m╔╝    \033[38;5;125m██\033[37m║   \033[38;5;125m██\033[37m║   \033[38;5;125m██\033[37m║\033[38;5;125m██\033[37m║   \033[38;5;125m██\033[37m║\033[38;5;125m██\033[37m║     \n")
			fmt.Printf(" \033[48;0;36m██\033[37m║   \033[48;0;36m██\033[37m║\033[48;0;36m██\033[37m╔═\033[48;0;36m██\033[37m╗    \033[38;5;125m██\033[37m║   \033[38;5;125m██\033[37m║   \033[38;5;125m██\033[37m║\033[38;5;125m██\033[37m║   \033[38;5;125m██\033[37m║\033[38;5;125m██\033[37m║     \n")
			fmt.Printf(" ╚\033[48;0;36m██████\033[37m╔╝\033[48;0;36m██\033[37m║  \033[48;0;36m██\033[37m╗   \033[38;5;125m██\033[37m║   ╚\033[38;5;125m██████\033[37m╔╝╚\033[38;5;125m██████\033[37m╔╝\033[38;5;125m███████\033[37m╗\n")
			fmt.Printf("  ╚═════╝ ╚═╝  ╚═╝   ╚═╝    ╚═════╝  ╚═════╝ ╚══════╝\033[0m\n\n")

			rand.Seed(time.Now().UTC().UnixNano())

			var ks []Key

			for i := 0; i < quantity; i++ {
				k := NewKey(length, noUpper, noLower, noNumber, noSymbol)
				ks = append(ks, k)
				switch {
				case int(k.strength) >= 0 && int(k.strength) < 60:
					color = LOW
				case int(k.strength) >= 60 && int(k.strength) < 90:
					color = MEDIUM
				case int(k.strength) >= 90:
					color = STRONG
				}
				fmt.Printf("\033[1mLENGTH\033[0m:%s%d\033[0m\n", color, k.length)
				fmt.Printf("\033[1mKEY%d\033[0m:%s%s\033[0m\n", i+1, color, k.value)
				fmt.Printf("\033[1mSTRENGTH\033[0m:%s%f\033[0m\n\n", color, k.strength)
			}

			if export {
				exportKeyInFile(ks)
				fmt.Printf("[\033[5;38;5;75mINFO\033[0m] Key(s) export in 'password.txt'\n")
			}

			return nil
		},
	}

	cli.VersionFlag = &cli.BoolFlag{
		Name:    "version",
		Aliases: []string{"V"},
		Usage:   "print only the version",
	}

	if err := app.Run(os.Args); err != nil {
		log.Fatal(err)
	}
}
